[DataLoader] DataLoader now automatically apply sharding to DataPipes

VitalyFedyunin opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #78631

This code assumes that every rank in distributed setup would have an equal amount of workers. I will drop this assumption in follow-up PR.

VitalyFedyunin committed:
[DataLoader] DataLoader now automatically apply sharding to DataPipes

[ghstack-poisoned]

VitalyFedyunin committed:
Update on "[DataLoader] DataLoader now automatically apply sharding to DataPipes"


This code assumes that every rank in distributed setup would have an equal amount of workers. I will drop this assumption in follow-up PR.

[ghstack-poisoned]

ejguan commented on torch/utils/data/dataloader.py:
```diff
@@ -90,6 +93,12 @@ def __iter__(self):
 
 
+def _sharding_worker_init_fn(worker_init_fn, worker_id):
+    if worker_init_fn is not None:
+        worker_init_fn(worker_id)
+    torch.utils.data.backward_compatibility.private_worker_init_fn(
+        worker_id)
```
We should call `private_worker_init_fn` before custom `worker_init_fn`, just in case users want to override the default behavior.

NivekT reviewed (commented):
It makes sense to me that sharding should get automatically applied so that users do not have to do it. The only case where it can raise an error is when `apply_sharding` is called twice but I can't think of a reason why it would be unless the users choose to call it.

nit: it will be helpful to add a note to the docstring of `ShardingFilter` before the args:
```python
Note:
        ``apply_sharding`` is automatically called while using ``torch.utils.data.DataLoader``, you do not need to call it again.
```

NivekT commented on torch/utils/data/backward_compatibility.py:
```diff
@@ -1,8 +1,20 @@
+                  " as DataLoader automatically calls it in every worker")
+
+
+def private_worker_init_fn(worker_id):
```
IIUC, the only behavior is `torch.utils.data.graph_settings.apply_sharding()`. Can that be overridden? I think if users call `apply_sharding` again, it will just raise an exception.

VitalyFedyunin commented on torch/utils/data/backward_compatibility.py:
```diff
@@ -1,8 +1,20 @@
 
 
 def worker_init_fn(worker_id):
+    warnings.warn("Usage of backward_compatibility.worker_init_fn is deprecated"
+                  " as DataLoader automatically calls it in every worker")
```
I want to avoid situations when users manually pass backward_compatibility.worker_init_fn as argument for classical DataLoader

VitalyFedyunin commented on torch/utils/data/dataloader_experimental.py:
```diff
@@ -59,12 +57,6 @@ def clean_me(thread, req_queue, res_queue):
-    torch.utils.data.backward_compatibility.worker_init_fn(
-        worker_id)
-
 class DataLoader2:
```
Eventually. Out of scope for this PR, as we want to cherry-pick it into release branch, so we keep it minimal.

VitalyFedyunin committed:
Update on "[DataLoader] DataLoader now automatically apply sharding to DataPipes"


This code assumes that every rank in distributed setup would have an equal amount of workers. I will drop this assumption in follow-up PR.

[ghstack-poisoned]

VitalyFedyunin committed:
Update on "[DataLoader] DataLoader now automatically apply sharding to DataPipes"


This code assumes that every rank in distributed setup would have an equal amount of workers. I will drop this assumption in follow-up PR.

[ghstack-poisoned]

VitalyFedyunin commented:
> It makes sense to me that sharding should get automatically applied so that users do not have to do it. The only case where it can raise an error is when `apply_sharding` is called twice but I can't think of a reason why it would be unless the users choose to call it.
> 
> nit: it will be helpful to add a note to the docstring of `ShardingFilter` before the args:
> 
> ```python
> Note:
>         ``apply_sharding`` is automatically called while using ``torch.utils.data.DataLoader``, you do not need to call it again.
> ```

Users are not supposed to call `apply_sharding` manually anyway, not sure I want to mention it.

VitalyFedyunin committed:
Update on "[DataLoader] DataLoader now automatically apply sharding to DataPipes"


This code assumes that every rank in distributed setup would have an equal amount of workers. I will drop this assumption in follow-up PR.

[ghstack-poisoned]

ejguan reviewed (approved):
LGTM, Thank you.

ejguan commented on torch/utils/data/dataloader.py:
```diff
@@ -90,6 +93,18 @@ def __iter__(self):
+    info = torch.utils.data.get_worker_info()
+    total_workers = info.num_workers
+    datapipe = info.dataset
+    if dist.is_available() and c10d.is_initialized():
```
nit: Why not use `dist.is_initialized()`?

VitalyFedyunin replied:
Because I had no idea it is exists =)

VitalyFedyunin committed:
Update on "[DataLoader] DataLoader now automatically apply sharding to DataPipes"


This code assumes that every rank in distributed setup would have an equal amount of workers. I will drop this assumption in follow-up PR.

[ghstack-poisoned]

VitalyFedyunin commented:
@pytorchbot merge -a

NivekT commented:
> @pytorchbot merge -a

@VitalyFedyunin `-a` doesn't work, it got reverted. `-g` will try to merge as soon as the builds (not test) are done.

ejguan reviewed (commented):
I just realize we might miss one use case. Please correct me if I am wrong. If we run distributed training with a single worker process, the `datapipe` won't be sharded since `worker_init_fn` is never executed. I think we need to add the corresponding logic in [_SingleProcessDataLoaderIter](https://github.com/pytorch/pytorch/blob/22b10873f375cb0d582f4567525238f716c430ca/torch/utils/data/dataloader.py#L627-L634)

VitalyFedyunin commented:
Indeep with num_workers = 0 it will not call the sharding function. Fixing it as follow-up.
